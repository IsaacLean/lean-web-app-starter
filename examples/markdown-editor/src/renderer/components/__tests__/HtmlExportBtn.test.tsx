import { render, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { useReducer, useRef } from 'react'

import { INITIAL_STATE, errorMessageReducer } from '../../errorMessageReducer'
import HtmlExportBtn from '../HtmlExportBtn'

function TestComponent() {
    const [, errorMessageDispatch] = useReducer(
        errorMessageReducer,
        INITIAL_STATE,
    )

    const refPreview = useRef<HTMLElement>(null)

    return (
        <>
            <HtmlExportBtn
                refPreview={refPreview}
                errorMessageDispatch={errorMessageDispatch}
            />
            <article ref={refPreview} />
        </>
    )
}

beforeEach(() => {
    // Mock the API
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    window.api = {
        showHtmlExportDialog: jest.fn(),
    }
})

/**
 * Setup user-event and render the component.
 */
const setupTest = () => {
    const user = userEvent.setup()
    const { asFragment } = render(<TestComponent />)

    return { asFragment, user }
}

test('HtmlExportBtn component matches snapshot', () => {
    const { asFragment } = setupTest()
    expect(asFragment()).toMatchSnapshot()
})

test('HtmlExportBtn component initiates html export process when clicked', async () => {
    const { user } = setupTest()

    expect(
        (
            window.api.showHtmlExportDialog as jest.MockedFunction<
                typeof window.api.showHtmlExportDialog
            >
        ).mock.calls,
    ).toHaveLength(0)

    await user.click(screen.getByRole('button'))

    expect(
        (
            window.api.showHtmlExportDialog as jest.MockedFunction<
                typeof window.api.showHtmlExportDialog
            >
        ).mock.calls,
    ).toHaveLength(1)
})
