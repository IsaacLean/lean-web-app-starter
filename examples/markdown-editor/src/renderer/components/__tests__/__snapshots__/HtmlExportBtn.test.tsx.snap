// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HtmlExportBtn component matches snapshot 1`] = `
<DocumentFragment>
  <button
    class="btn"
  >
    Export as HTML
  </button>
  <article />
</DocumentFragment>
`;
